cdk/apps/ica_credentials/secrets/runtime/jwt_producer/lambda_function.py

```diff
@@ -0,0 +1,137 @@
+import boto3
+import json
+import os
+import urllib3
+import traceback
+
+
+def main(ev, _):
+    """
+    Uses a master API key secret to generate new JWTs.
+    """
+    arn = ev["SecretId"]
+    tok = ev["ClientRequestToken"]
+    step = ev["Step"]
+
+    print(f"Starting step '{step}' of Secret Rotation for secret {arn}")
+
+    master_arn = os.environ["MASTER_ARN"]
+    ica_base_url = os.environ["ICA_BASE_URL"]
+
+    sm_client = boto3.client("secretsmanager")
+
+    try:
+        master_resp = sm_client.get_secret_value(SecretId=master_arn)
+
+        master_val = master_resp["SecretString"]
+
+    except Exception as e:
+        print("Error fetching master secret")
+        print(traceback.format_exc())
+        raise e
+
+    # make sure the version is staged correctly
+    try:
+        metadata = sm_client.describe_secret(SecretId=arn)
+
+        if not metadata["RotationEnabled"]:
+            raise ValueError(f"Secret {arn} is not enabled for rotation")
+
+        versions = metadata["VersionIdsToStages"]
+
+        if tok not in versions:
+            raise ValueError(
+                f"Secret version {tok} has no stage for rotation of secret {arn}"
+            )
+
+        if "AWSCURRENT" in versions[tok]:
+            return
+        elif "AWSPENDING" not in versions[tok]:
+            raise ValueError(
+                f"Secret version ${tok} not set as AWSPENDING for rotation of secret {arn}"
+            )
+    except Exception as e:
+        print("Error with state of secrets steps")
+        print(traceback.format_exc())
+        raise e
+
+    try:
+        if step == "createSecret":
+            create_secret(sm_client, master_val, ica_base_url, arn, tok)
+        elif step == "setSecret":
+            pass
+            # set_secret(service_client, arn, token)
+        elif step == "testSecret":
+            pass
+        elif step == "finishSecret":
+            finish_secret(sm_client, arn, tok)
+        else:
+            raise ValueError("Invalid step parameter")
+    except Exception as e:
+        print(f"Error with step '{step}'")
+        print(traceback.format_exc())
+        raise e
+
+
+def create_secret(
+    client, master_val, ica_base_url, arn, tok
+):
+    """
+    Create a new JWT via the ICA APIs
+    """
+    try:
+        # if get works here then we have a value and we should *not* create
+        client.get_secret_value(SecretId=arn, VersionId=tok, VersionStage="AWSPENDING")
+    except:
+        http = urllib3.PoolManager()
+
+        r = http.request(
+            "POST",
+            f"{ica_base_url}/v1/tokens",
+            headers={
+                "Accept": "application/json",
+                "X-API-Key": master_val,
+            },
```

Comment:
Might need to be per-project-context though right? This would just create a 'personal-token' rather than a project specific token.  

Use `data = { "cid": project_id }`  

This shouldn't fail, if it does OR you would like to check if the project_id var is valid, then this might become a two-step process:

1. Create a "personal" token without the project id (as has been done above), use this token to validate the project id against 
https://ica-docs.readme.io/reference#listprojects
2. Use the new "personal" token in conjunction with the api key to create a project specific token:  
Example:  https://github.com/umccr/ica-ica-lazy/blob/main/scripts/ica-add-access-token#L160-L172

Reply:
Isn't this JWT to be used across all projects though? I thought this was like the service user for the whole portal?

Reply:
@reisingerf when you make your service user API keys - do you make one for dev and one for prod - and at that point build in the projects it can access? Or would you  like the JWT generator to assume a API key with highest of permissions and have it restrict down?

Reply:
This is the JWT of the current setup

```
{
  "sub": "bc99b89c-3bb7-334b-80d1-20ef9e65f0b0",
  "iss": "https://platform.login.illumina.com",
  "acl": [
    "cid:dc8e6ba9-b744-437b-b070-4cf014694b3d",
    "tid:YXdzLXVzLXBsYXRmb3JtOjEwMDAwNTM3OjBiYTU5YWUxLWZkYWUtNDNiYS1hM2I1LTRkMzY3YTQzYWJkNQ"
  ],
  "gty": "api_key",
  "tid": "YXdzLXVzLXBsYXRmb3JtOjEwMDAwNTM3OjBiYTU5YWUxLWZkYWUtNDNiYS1hM2I1LTRkMzY3YTQzYWJkNQ",
  "aud": "iap-aps2",
  "nbf": 1628670485,
  "mem": {
    "cid:dc8e6ba9-b744-437b-b070-4cf014694b3d": "*"
  },
  "azp": "iap-aps2",
  "scope": "GDS.FOLDERS.DELETE,GMS.ANALYSISRUNS.CREATE,WES.SIGNALS.UPDATE,GSS.ANALYSISVERSIONS.UPDATE,GSS.LI,GDS.FOLDERS.READ,GMS.ANALYSISDEFINITIONS.GRANT,WES.SIGNALS.READ,GMS.ANALYSISVERSIONDEFINITIONS.CREATE,IMS.INSTRUMENTS.UPDATE,WES.WORKFLOWS.CREATE,GSS.ANALYSISDATASETTYPE.READ,WES.VERSIONS.DELETE,GSS.SEQUENCINGRUNS.READ,GSS.ANALYSISRUNS.READ,BSSH.RUNS.READ,GSS.ANALYSISDATASETTYPES.CREATE,GSS.LIBRARYPREPKITS.READ,GSS.LIBRARYPREPKITS.CREATE,GSS.ANALYSISVERSIONS.READ,GSS.SAMPLES.GRANT,GDS.VOLUMES.GRANT,TES.RUNS.DELETE,GDS.VOLUMES.CREATE,GSS.ANALYSISDATASETS.CREATE,IMS.ASSETVERSIONS.READ,GSS.GENOMES.READ,TES.TASKS.UPDATE,GMS.BEADCHIPS.READ,GSS.ANALYSISDATASETTYPES.UPDATE,GSS.INDEXADAPTERKITS.READ,IMS.ASSETS.READ,GMS.PRODUCTS.DELETE,GDS.FOLDERS.GRANT,GSS.ANALYSISVERSIONS.CREATE,GMS.SAMPLES.GRANT,GSS.ANALYSISDATASET.CREATE,GDS.FILES.UPDATE,GSS.ANALYSISRUNS.GRANT,WES.RUNS.UPDATE,GSS.LIBRARIES.CREATE,GMS.MANIFESTS.DELETE,GMS.MANIFESTS.GRANT,IMS.INSTRUMENTS.READ,GMS.ANALYSISCONFIGURATIONS.GRANT,GSS.SAMPLES.READ,GSS.SEQUENCINGRUNS.GRANT,GSS.INDEXADAPTERKITS.GRANT,TES.VERSIONS.GRANT,GSS.LIBRARYPOOLS.GRANT,GSS.INDEXADAPTERKITS.CREATE,GDS.FOLDERS.ARCHIVE,GDS.VOLUMES.DELETE,GSS.ANALYSISDATASETS.DELETE,GMS.ANALYSISFILES.READ,WES.VERSIONS.READ,GSS.ANALYSISDATASET.UPDATE,GSS.ANALYSISDEFINITIONS.READ,IMS.INSTRUMENTS.GRANT,WES.VERSIONS.GRANT,GMS.BEADCHIPS.GRANT,GSS.ANALYSISDATASETTYPES.GRANT,ENS.SUBSCRIPTIONS.MANAGE,GSS.LIBRARYPOOLS.DELETE,GSS.ANALYSISDEFINITIONS.CREATE,GSS.SAMPLES.CREATE,TES.VERSIONS.READ,GSS.ANALYSISDATASETS.READ,TES.VERSIONS.CREATE,GMS.ANALYSISVERSIONDEFINITIONS.READ,GDS.FILES.DELETE,GSS.GENOMES.GRANT,WES.RUNS.GRANT,GSS.ANALYSISDATASETTYPE.GRANT,GMS.SAMPLES.CREATE,GSS.SEQUENCINGRUNS.UPDATE,DCS.USAGES.READ,GMS.ANALYSISVERSIONDEFINITIONS.GRANT,GSS.ANALYSISVERSIONS.GRANT,TES.RUNS.CREATE,WES.SIGNALS.GRANT,GDS.FILES.READ,GMS.BEADCHIPS.DELETE,GSS.ANALYSISVERSIONS.DELETE,GSS.LIBRARIES.READ,GMS.MANIFESTS.UPDATE,GSS.GENOMES.CREATE,WES.RUNS.DELETE,IMS.INSTRUMENTS.DELETE,GMS.ANALYSISCONFIGURATIONS.READ,TES.TASKS.READ,GMS.MANIFESTS.CREATE,GMS.RUNS.UPDATE,GMS.ANALYSISRUNS.GRANT,WES.SIGNALS.DELETE,GDS.FOLDERS.UPDATE,GMS.ANALYSISDEFINITIONS.CREATE,GSS.ANALYSISDATASETTYPE.CREATE,GSS.ANALYSISDATASETTYPES.READ,TES.TASKS.DELETE,GSS.LIBRARYPOOLS.CREATE,GSS.ANALYSISRUNS.CREATE,GDS.VOLUMES.ARCHIVE,TES.RUNS.GRANT,GMS.ANALYSISCONFIGURATIONS.CREATE,GDS.FILES.DOWNLOAD,GMS.RUNS.GRANT,GMS.ANALYSISRUNS.READ,GMS.PRODUCTS.CREATE,GSS.INDEXADAPTERKITS.UPDATE,WES.VERSIONS.CREATE,WES.WORKFLOWS.READ,GSS.LIBRARIES.GRANT,GMS.ANALYSISFILES.GRANT,TES.TASKS.GRANT,GSS.LIBRARYPREPKITS.DELETE,GMS.PRODUCTS.GRANT,WES.RUNS.CREATE,TES.VERSIONS.UPDATE,GSS.LIBRARIES.UPDATE,GMS.PRODUCTS.READ,GSS.LIBRARYPREPKITS.GRANT,GSS.ANALYSISDEFINITIONS.GRANT,GMS.ANALYSISDEFINITIONS.READ,GSS.LIBRARYPOOLS.UPDATE,GMS.BEADCHIPS.UPDATE,GMS.RUNS.READ,GMS.BEADCHIPS.CREATE,GSS.ANALYSISRUNS.DELETE,GSS.ANALYSISDATASETTYPE.UPDATE,WES.WORKFLOWS.UPDATE,GSS.ANALYSISRUNS.UPDATE,GSS.SEQUENCINGRUNS.DELETE,TES.RUNS.READ,GDS.FOLDERS.CREATE,GDS.VOLUMES.UPDATE,GMS.ANALYSISFILES.CREATE,GSS.ANALYSISDATASETS.UPDATE,GSS.SAMPLES.UPDATE,GDS.FILES.ARCHIVE,GDS.VOLUMES.READ,GSS.ANALYSISDEFINITIONS.DELETE,IMS.INSTRUMENTS.CREATE,TES.VERSIONS.DELETE,GSS.ANALYSISDATASET.READ,GDS.FILES.CREATE,GMS.MANIFESTS.READ,GSS.SAMPLES.DELETE,TES.RUNS.UPDATE,GMS.SAMPLES.READ,GSS.ANALYSISDEFINITIONS.UPDATE,WES.WORKFLOWS.DELETE,GSS.INDEXADAPTERKITS.DELETE,GSS.LIBRARYPOOLS.READ,GSS.SEQUENCINGRUNS.CREATE,GSS.ANALYSISDATASET.GRANT,GSS.GENOMES.UPDATE,GMS.PRODUCTS.UPDATE,GMS.RUNS.CREATE,TES.TASKS.CREATE,WES.WORKFLOWS.GRANT,WES.VERSIONS.UPDATE,GSS.ANALYSISDATASETS.GRANT,GSS.LIBRARYPREPKITS.UPDATE,WES.RUNS.READ,WES.SIGNALS.CREATE,GSS.LIBRARIES.DELETE",
  "tns": "umccr-beta",
  "exp": 1637310485,
  "iat": 1628670485,
  "jti": ""
}

Reply:
This is my JWT is produced from my personal API key.. certainly different around the "mem" claim

```
{
  "sub": "4c6d4258-e621-3f43-9dde-1767d5255c4f",
  "iss": "https://platform.login.illumina.com",
  "acl": [
    "tid:YXdzLXVzLXBsYXRmb3JtOjEwMDAwNTM3OjBiYTU5YWUxLWZkYWUtNDNiYS1hM2I1LTRkMzY3YTQzYWJkNQ",
    "uid:4c6d4258-e621-3f43-9dde-1767d5255c4f"
  ],
  "gty": "api_key",
  "tid": "YXdzLXVzLXBsYXRmb3JtOjEwMDAwNTM3OjBiYTU5YWUxLWZkYWUtNDNiYS1hM2I1LTRkMzY3YTQzYWJkNQ",
  "aud": "iap-aps2",
  "nbf": 1630288521,
  "mem": {
    "wid:5a7cc0c2-2e9c-3438-b145-ab2e72f0c195": "0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,112,113,114,115,116,117,118,119,120,121,122,123,124,125,126,127,128,129,130,131,132,133,134,135,136,137,138,139,140,141,142,143,144,145,146,147,148,149,150,151,152,153",
    "uid:4c6d4258-e621-3f43-9dde-1767d5255c4f": "*"
  },
  "azp": "iap-aps2",
  "scope": "GDS.FOLDERS.DELETE,GMS.ANALYSISRUNS.CREATE,WES.SIGNALS.UPDATE,GSS.ANALYSISVERSIONS.UPDATE,GDS.FOLDERS.READ,GMS.ANALYSISDEFINITIONS.GRANT,WES.SIGNALS.READ,GMS.ANALYSISVERSIONDEFINITIONS.CREATE,IMS.INSTRUMENTS.UPDATE,WES.WORKFLOWS.CREATE,WES.VERSIONS.DELETE,GSS.SEQUENCINGRUNS.READ,GSS.ANALYSISRUNS.READ,BSSH.RUNS.READ,GSS.LIBRARYPREPKITS.READ,GSS.LIBRARYPREPKITS.CREATE,GSS.ANALYSISVERSIONS.READ,GSS.SAMPLES.GRANT,GDS.VOLUMES.GRANT,TES.RUNS.DELETE,GDS.VOLUMES.CREATE,GSS.ANALYSISDATASETS.CREATE,IMS.ASSETVERSIONS.READ,GSS.GENOMES.READ,TES.TASKS.UPDATE,GMS.BEADCHIPS.READ,GSS.INDEXADAPTERKITS.READ,IMS.ASSETS.READ,GMS.PRODUCTS.DELETE,GDS.FOLDERS.GRANT,GSS.ANALYSISVERSIONS.CREATE,GMS.SAMPLES.GRANT,GDS.FILES.UPDATE,GSS.ANALYSISRUNS.GRANT,WES.RUNS.UPDATE,GSS.LIBRARIES.CREATE,GMS.MANIFESTS.DELETE,GMS.MANIFESTS.GRANT,IMS.INSTRUMENTS.READ,GMS.ANALYSISCONFIGURATIONS.GRANT,GSS.SAMPLES.READ,GSS.SEQUENCINGRUNS.GRANT,GSS.INDEXADAPTERKITS.GRANT,TES.VERSIONS.GRANT,GSS.LIBRARYPOOLS.GRANT,GSS.INDEXADAPTERKITS.CREATE,GDS.FOLDERS.ARCHIVE,GDS.VOLUMES.DELETE,GSS.ANALYSISDATASETS.DELETE,GMS.ANALYSISFILES.READ,WES.VERSIONS.READ,GSS.ANALYSISDEFINITIONS.READ,IMS.INSTRUMENTS.GRANT,WES.VERSIONS.GRANT,GMS.BEADCHIPS.GRANT,ENS.SUBSCRIPTIONS.MANAGE,GSS.LIBRARYPOOLS.DELETE,GSS.ANALYSISDEFINITIONS.CREATE,GSS.SAMPLES.CREATE,TES.VERSIONS.READ,GSS.ANALYSISDATASETS.READ,TES.VERSIONS.CREATE,GMS.ANALYSISVERSIONDEFINITIONS.READ,GDS.FILES.DELETE,GSS.GENOMES.GRANT,WES.RUNS.GRANT,GMS.SAMPLES.CREATE,GSS.SEQUENCINGRUNS.UPDATE,DCS.USAGES.READ,GMS.ANALYSISVERSIONDEFINITIONS.GRANT,GSS.ANALYSISVERSIONS.GRANT,TES.RUNS.CREATE,WES.SIGNALS.GRANT,GDS.FILES.READ,GMS.BEADCHIPS.DELETE,GSS.ANALYSISVERSIONS.DELETE,GSS.LIBRARIES.READ,GMS.MANIFESTS.UPDATE,GSS.GENOMES.CREATE,WES.RUNS.DELETE,IMS.INSTRUMENTS.DELETE,GMS.ANALYSISCONFIGURATIONS.READ,TES.TASKS.READ,GMS.MANIFESTS.CREATE,GMS.RUNS.UPDATE,GMS.ANALYSISRUNS.GRANT,WES.SIGNALS.DELETE,GDS.FOLDERS.UPDATE,GMS.ANALYSISDEFINITIONS.CREATE,TES.TASKS.DELETE,GSS.LIBRARYPOOLS.CREATE,GSS.ANALYSISRUNS.CREATE,GDS.VOLUMES.ARCHIVE,TES.RUNS.GRANT,GMS.ANALYSISCONFIGURATIONS.CREATE,GDS.FILES.DOWNLOAD,GMS.RUNS.GRANT,GMS.ANALYSISRUNS.READ,GMS.PRODUCTS.CREATE,GSS.INDEXADAPTERKITS.UPDATE,WES.VERSIONS.CREATE,WES.WORKFLOWS.READ,GSS.LIBRARIES.GRANT,GMS.ANALYSISFILES.GRANT,TES.TASKS.GRANT,GSS.LIBRARYPREPKITS.DELETE,GMS.PRODUCTS.GRANT,WES.RUNS.CREATE,TES.VERSIONS.UPDATE,GSS.LIBRARIES.UPDATE,GMS.PRODUCTS.READ,GSS.LIBRARYPREPKITS.GRANT,GSS.ANALYSISDEFINITIONS.GRANT,GMS.ANALYSISDEFINITIONS.READ,GSS.LIBRARYPOOLS.UPDATE,GMS.BEADCHIPS.UPDATE,GMS.RUNS.READ,GMS.BEADCHIPS.CREATE,GSS.ANALYSISRUNS.DELETE,WES.WORKFLOWS.UPDATE,GSS.ANALYSISRUNS.UPDATE,GSS.SEQUENCINGRUNS.DELETE,TES.RUNS.READ,GDS.FOLDERS.CREATE,GDS.VOLUMES.UPDATE,GMS.ANALYSISFILES.CREATE,GSS.ANALYSISDATASETS.UPDATE,GSS.SAMPLES.UPDATE,GDS.FILES.ARCHIVE,GDS.VOLUMES.READ,GSS.ANALYSISDEFINITIONS.DELETE,IMS.INSTRUMENTS.CREATE,TES.VERSIONS.DELETE,GDS.FILES.CREATE,GMS.MANIFESTS.READ,GSS.SAMPLES.DELETE,TES.RUNS.UPDATE,GMS.SAMPLES.READ,GSS.ANALYSISDEFINITIONS.UPDATE,WES.WORKFLOWS.DELETE,GSS.INDEXADAPTERKITS.DELETE,GSS.LIBRARYPOOLS.READ,GSS.SEQUENCINGRUNS.CREATE,GSS.GENOMES.UPDATE,GMS.PRODUCTS.UPDATE,GMS.RUNS.CREATE,TES.TASKS.CREATE,WES.WORKFLOWS.GRANT,WES.VERSIONS.UPDATE,GSS.ANALYSISDATASETS.GRANT,GSS.LIBRARYPREPKITS.UPDATE,WES.RUNS.READ,WES.SIGNALS.CREATE,GSS.LIBRARIES.DELETE",
  "tns": "umccr-beta",
  "exp": 1638928521,
  "iat": 1630288521,
  "jti": ""
}
```

Reply:
So @alexiswl it does look like the current one is project specific to dev (with a match cid: claim). I'll convert to the project specific call then

Reply:
> "mem" claim
Yep looks like it has a workgroup context (which I don't know why that would be?)

The `acl` attribute is different too. 

Also one thing to be wary of... 
https://github.com/umccr-illumina/stratus/issues/151

Reply:
> Or would you like the JWT generator to assume a API key with highest of permissions and have it restrict down to named project ids?

At the moment, the API key generates a token based on the users scope in the project (rather they're a union of all a user's workgroups / privileges for a given project) - https://github.com/umccr/ica-ica-lazy/wiki/Getting_Started#saving-your-api-key

Reply:
So for now it doesn't really matter on the permissions of the api key because the token will be generated with the full level of permissions that ica user has for that project (unless the scopes are used)

Reply:
> > "mem" claim
> > Yep looks like it has a workgroup context (which I don't know why that would be?)
> 
> The `acl` attribute is different too.
> 
> Also one thing to be wary of...
> [umccr-illumina/stratus#151](https://github.com/umccr-illumina/stratus/issues/151)

@alexiswl 
How many seconds have you experienced worst case?? I can certainly get the actual secret rotation code to pause for 10 seconds after the JWT is issued before it becomes available to others? Even minutes are possible.. beyond 10 mins I will struggle..

Reply:
Token are now generated with project permissions.
There is a 30 second delay after token generation before it is saved for use.